// @crypto.com/facilitator-client/dist/integrations/facilitator.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contract = exports.X402EventType = exports.CronosNetwork = exports.Scheme = void 0;
/**
 * X402 payment scheme type.
 *
 * Currently, the facilitator supports only `exact` payments, meaning the
 * payment must match the required amount exactly rather than range-based.
 */
var Scheme;
(function (Scheme) {
    Scheme["Exact"] = "exact";
})(Scheme || (exports.Scheme = Scheme = {}));
/**
 * Supported Cronos EVM networks for X402 flows.
 *
 * These map directly to `NETWORK_REGISTRY` entries.
 */
var CronosNetwork;
(function (CronosNetwork) {
    /** Cronos EVM Mainnet */
    CronosNetwork["CronosMainnet"] = "cronos-mainnet";
    /** Cronos EVM Testnet (chainId: 338) */
    CronosNetwork["CronosTestnet"] = "cronos-testnet";
})(CronosNetwork || (exports.CronosNetwork = CronosNetwork = {}));
/**
 * Event types emitted during settlement.
 */
var X402EventType;
(function (X402EventType) {
    /** Payment successfully settled on-chain. */
    X402EventType["PaymentSettled"] = "payment.settled";
    /** Payment verification or execution failed. */
    X402EventType["PaymentFailed"] = "payment.failed";
})(X402EventType || (exports.X402EventType = X402EventType = {}));
/**
 * Supported ERC-20 contract addresses accepted by the Cronos X402 facilitator.
 *
 * These contracts implement the **EIP-3009 TransferWithAuthorization**
 * interface and are validated by the facilitator during verification
 * and settlement.
 *
 * The SDK restricts payments to this enum to prevent unsupported
 * or unsafe token usage.
 */
var Contract;
(function (Contract) {
    /**
     * USDCe contract address on Cronos mainnet.
     *
     * Implements EIP-3009 and is accepted by the production facilitator.
     */
    Contract["USDCe"] = "0xf951eC28187D9E5Ca673Da8FE6757E6f0Be5F77C";
    /**
     * USDCe contract address on Cronos testnet.
     *
     * Intended for development and testing environments only.
     */
    Contract["DevUSDCe"] = "0xc01efAaF7C5C61bEbFAeb358E1161b537b8bC0e0";
})(Contract || (exports.Contract = Contract = {}));
